vms/platformvm/state/state.go

```diff
@@ -944,74 +966,168 @@ func (s *state) ValidatorSet(subnetID ids.ID, vdrs validators.Set) error {
 	return nil
 }
 
-func (s *state) GetValidatorWeightDiffs(height uint64, subnetID ids.ID) (map[ids.NodeID]*ValidatorWeightDiff, error) {
-	prefixStruct := heightWithSubnet{
-		Height:   height,
-		SubnetID: subnetID,
-	}
-	prefixBytes, err := blocks.GenesisCodec.Marshal(blocks.Version, prefixStruct)
-	if err != nil {
-		return nil, err
-	}
-	prefixStr := string(prefixBytes)
+func (s *state) ApplyValidatorWeightDiffs(
+	ctx context.Context,
+	validators map[ids.NodeID]*validators.GetValidatorOutput,
+	startHeight uint64,
+	endHeight uint64,
+	subnetID ids.ID,
+) error {
+	diffIter := s.flatValidatorWeightDiffsDB.NewIteratorWithStartAndPrefix(
+		getStartWeightKey(subnetID, startHeight),
+		subnetID[:],
+	)
+	defer diffIter.Release()
 
-	if weightDiffs, ok := s.validatorWeightDiffsCache.Get(prefixStr); ok {
-		return weightDiffs, nil
-	}
+	prevHeight := startHeight + 1
+	// TODO: Remove the index continuity checks once we are guaranteed nodes can
+	// not rollback to not support the new indexing mechanism.
+	for diffIter.Next() && s.indexedHeights != nil && s.indexedHeights.LowerBound <= endHeight {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
 
-	rawDiffDB := prefixdb.New(prefixBytes, s.validatorWeightDiffsDB)
-	diffDB := linkeddb.NewDefault(rawDiffDB)
-	diffIter := diffDB.NewIterator()
-	defer diffIter.Release()
+		_, parsedHeight, nodeID, err := parseWeightKey(diffIter.Key())
+		if err != nil {
+			return err
+		}
+		// If the parsedHeight is less than our target endHeight, then we have
+		// fully processed the diffs from startHeight through endHeight.
+		if parsedHeight < endHeight {
+			return diffIter.Error()
+		}
 
-	weightDiffs := make(map[ids.NodeID]*ValidatorWeightDiff)
-	for diffIter.Next() {
-		nodeID, err := ids.ToNodeID(diffIter.Key())
+		prevHeight = parsedHeight
+
+		weightDiff, err := parseWeightValue(diffIter.Value())
 		if err != nil {
-			return nil, err
+			return err
+		}
+
+		if err := applyWeightDiff(validators, nodeID, weightDiff); err != nil {
+			return err
+		}
+	}
```

Comment:
Do we need to check diffIter.Error here?

Reply:
We return the iterator error at the bottom of the function.

Reply:
nit: Usually, I'd return from error ASAP